mumd/src/audio/sound_effects.rs

```diff
@@ -0,0 +1,216 @@
+            "server_disconnect" => Ok(NotificationEvents::ServerDisconnect),
+            "user_connected" => Ok(NotificationEvents::UserConnected),
+            "user_disconnected" => Ok(NotificationEvents::UserDisconnected),
+            "user_joined_channel" => Ok(NotificationEvents::UserJoinedChannel),
+            "user_left_channel" => Ok(NotificationEvents::UserLeftChannel),
+            "mute" => Ok(NotificationEvents::Mute),
+            "unmute" => Ok(NotificationEvents::Unmute),
+            "deafen" => Ok(NotificationEvents::Deafen),
+            "undeafen" => Ok(NotificationEvents::Undeafen),
+            _ => {
+                warn!("Unknown notification event '{}' in config", s);
+                Err(())
+            }
+        }
+    }
+}
+
+/// Loads files into an "event -> data"-map, with support for overriding
+/// specific events with another sound file.
+pub fn load_sound_effects(overrides: &[SoundEffect], num_channels: usize) -> HashMap<NotificationEvents, Vec<f32>> {
+    let overrides: HashMap<_, _> = overrides
+        .iter()
+        .filter_map(|sound_effect| {
+            let (event, file) = (&sound_effect.event, &sound_effect.file);
+            if let Ok(event) = NotificationEvents::try_from(event.as_str()) {
+                Some((event, file))
+            } else {
+                None
+            }
+        })
+        .collect();
+
+    // Construct a hashmap that maps every [NotificationEvent] to a vector of
+    // plain floating point audio data with the global sample rate as a
+    // Vec<f32>. We do this by iterating over all [NotificationEvent]-variants
+    // and opening either the file passed as an override or the fallback sound
+    // effect (if omitted). We then use dasp to convert to the correct sample rate.
+    NotificationEvents::iter()
+        .map(|event| {
+            let file = overrides.get(&event);
+            // Try to open the file if overriden, otherwise use the default sound effect.
+            let (data, kind) = file
+                .and_then(|file| {
+                    // Try to get the file kind from the extension.
+                    let kind = file
+                        .split('.')
+                        .last()
+                        .and_then(|ext| AudioFileKind::try_from(ext).ok())?;
+                    Some((get_sfx(file), kind))
+                })
+                .unwrap_or_else(|| (get_default_sfx(), AudioFileKind::Wav));
+            // Unpack the samples.
+            let (samples, spec) = unpack_audio(data, kind);
+            // If the audio is mono (single channel), pad every sample with
+            // itself, since we later assume that audio is stored interleaved as
+            // LRLRLR (or RLRLRL). Without this, mono audio would be played in
+            // double speed.
+            let iter: Box<dyn Iterator<Item = f32>> = match spec.channels {
+                1 => Box::new(samples.into_iter().flat_map(|e| vec![e, e])),
```

Comment:
This probably doesn't need to be a vec since arrays implement `IntoIterator` in rust 1.53 and later.